apps/labs/posts/scipy-array-api.md

```diff
@@ -0,0 +1,441 @@
+    xp = array_namespace(x)
+
+    if is_numpy(xp):
+        return pocketfft_func(x, n=n, axis=axis, norm=norm,
+                              overwrite_x=overwrite_x, workers=workers, plan=plan)
+
+    norm = _validate_fft_args(workers, plan, norm)
+    if hasattr(xp, 'fft'):
+        xp_func = getattr(xp.fft, func_str)
+        return xp_func(x, n=n, axis=axis, norm=norm)
+
+    x = np.asarray(x)
+    y = pocketfft_func(x, n=n, axis=axis, norm=norm)
+    return xp.asarray(y)
+
+def fft(x, n=None, axis=-1, norm=None,
+        overwrite_x=False, workers=None, *, plan=None):
+    return _execute_1D('fft', _pocketfft.fft, x, n=n, axis=axis, norm=norm,
+                       overwrite_x=overwrite_x, workers=workers, plan=plan)
+```
+
+After converting the production functions, we also need to convert the tests!
+
+We have the `@array_api_compatible` decorator, which parametrizes `xp` with the
+available backends that we have for testing.
+We have also written three new assertions - `assert_close`, `assert_equal` and `assert_less` -
+which incorporate checks for matching namespaces and dtypes.
+If any tests use results or input data from NumPy, we convert everything to the `xp` namespace before using
+these assertions.
+The assertions make use of functions from `xp` where possible which helps avoid device transfers.
+
+Other than that, we make similar changes to those in the production functions so that every test works
+with all of the backends we test with.
+Finally, for tests of functions which hit compiled code, we add the `@skip_if_array_api_gpu` decorator.
+
+Here is what that looks like for one of the Discrete Sine and Cosine Transforms tests:
+
+```diff
++ @skip_if_array_api_gpu
++ @array_api_compatible
+  @pytest.mark.parametrize("func", ['dct', 'dst', 'dctn', 'dstn'])
+  @pytest.mark.parametrize("type", [1, 2, 3, 4])
+  @pytest.mark.parametrize("norm", [None, 'backward', 'ortho', 'forward'])
+- def test_fftpack_equivalience(func, type, norm):
++ def test_fftpack_equivalience(func, type, norm, xp):
+      x = np.random.rand(8, 16)
++     fftpack_res = xp.asarray(getattr(fftpack, func)(x, type, norm=norm))
++     x = xp.asarray(x)
+      fft_res = getattr(fft, func)(x, type, norm=norm)
+-     fftpack_res = getattr(fftpack, func)(x, type, norm=norm)
+
+-     assert_allclose(fft_res, fftpack_res)
++     xp_assert_close(fft_res, fftpack_res)
+```
+
+## Machinery, Helpers and Utilities
+
+I'll talk a bit about implementation details now - feel free to skip this section if you are just here for
+an overview!
+
+When we call `array_namespace`, a little more is done than just returning the namespace.
+If the environment variable `SCIPY_ARRAY_API` is not set, `array_api_compat.numpy`
+(the compatible wrapper for `numpy`) is returned, regardless of the type of the input array.
+This allows us to retain all pre-array API behaviour unless the user has set the environment variable.
```

Comment:
I could clarify that it does this by sending all arrays down the `is_numpy` code path.